### ISSUE TITLE
Choices Dropdown Menu in API Documentation

### ISSUE THREAD

**Issue by alikhtag** [tiangolo/fastapi#246]
**Feature Request:**

It would be nice to document in the API, which what choices are available to the user as a drop-down menu in the UI. It helps prevent such documentation in the description\help of the parameter.  

**Context**
`flask_restplus` library is able to show choices in its Swagger UI, and is able to handle if incorrect choice is provided as an input by the user.  
![Selection_005](https://user-images.githubusercontent.com/43503284/58156014-6dffa880-7c75-11e9-9d35-3c32150a3596.png)

**Preferred Solution:**
Would like to see `choices` parameter added to possibly `fastapi.Query` class and possibly other classes that handle input from user. 
Would be awesome if it was something like this
```
@api.get('/get_countries')
def get_something(
    choice: str = Query('eu', choices=('eu,'us','cn,'ru'), description='foo bar')
):
    do_something(choice)
``` 

Would be nice to see this feature in `FastAPI`, as I/We would prefer to use `FastAPI` over `flask_restplus`.

**Comment by euri10** [comment]
after some digging it seems it works out of the box:

![Imgur](https://i.imgur.com/tRCaRXM.png)

```
from enum import Enum, IntEnum
from typing import Set

import uvicorn

from fastapi import FastAPI, Query
from pydantic import BaseModel

app = FastAPI()


@app.get('/get_countries')
async def get_countries(_q: str = Query("eu", enum=["eu", "us", "cn", "ru"])):
    return {"item": "yo"}


if __name__ == '__main__':
    uvicorn.run("test_swagger_dropdown:app", reload=True)
```

**Comment by tiangolo** [comment]
Thanks @euri10 for your help!

Later I'll add full `Enum` support for query values too, that way, apart from showing up in Swagger UI,  Pydantic validates the values given.

**Comment by alikhtag** [comment]
Thanks a lot! This will work! Can't wait for full support!

**Comment by tiangolo** [comment]
Hey @alikhtag, just to let you know, this is available now in the latest versions (actually for some time now).

Here are docs for how to use enums in paths, but the same applies to queries: https://fastapi.tiangolo.com/tutorial/path-params/#predefined-values

Here's your same example:

```Python
from fastapi import FastAPI
from enum import Enum


class Country(str, Enum):
    eu = "eu"
    us = "us"
    cn = "cn"
    ru = "ru"


app = FastAPI()


@app.get("/")
def get_something(country: Country = Country.eu):
    return {"country": country.value}
```

**Comment by alikhtag** [comment]
@tiangolo 
Thanks a lot!

**Comment by vstathok** [comment]
Hello, in the above setup, is there a way to select multiple values from the dropdown list?

**Comment by hvuhsg** [comment]
> 
> 
> Hello @tiangolo, in the above setup, is there a way to select multiple values from the dropdown list? Thanks a lot

I need this feature too

**Comment by ybressler** [comment]
Can confirm, the above method also works with enums that aren't explicitly string types:
```
import enum

class StatusEnum(enum.Enum):
    made_an_account = 'made_an_account'
    made_an_order = 'made_an_order'
```

@app.post("/set-status")
def set_user_status(id:int, status: StatusEnum = StatusEnum.made_an_account):
    print(f'You selected {status=}')
    pass

**Comment by sureshdsk** [comment]
@euri10 @tiangolo 

Looking for a way to show dynamic enums in swagger docs for different users based on their login session or by using their token or cookie or by any other means

```python
import uvicorn

from fastapi import FastAPI, Query

app = FastAPI()


@app.get('/get_countries')
async def get_countries(_q: str = Query("country", enum=["eu", "us", "cn", "ru"])):
## For user 1: enum=["eu", "us", "cn", "ru"]
## For user 2: enum=["in", "au", "uk"]
    return {"selected": _q}


if __name__ == '__main__':
    uvicorn.run("test_swagger_dropdown:app", reload=True)
```

Is there any solution to achieve this either by using fastapi or by any other libraries?

**Comment by sureshdsk** [comment]
@tiangolo I think I'm looking at extending open api spec.  Let me know if I'm going in the right direction

https://fastapi.tiangolo.com/advanced/extending-openapi/

**Comment by gokul427** [comment]
Looking for a way of displaying options in second parameter based on previous selection of first parameter.

**Comment by r3v4-onbloc** [comment]
Is there ways to apply dropdown menu at global scope? not per endpoints.
I've seen pic like below, but couldn't found ways to do it.

![image](https://user-images.githubusercontent.com/104744707/189833810-2d5aaa55-c4c7-4a40-a57d-cd763d70ebc1.png)

**Comment by iudeen** [comment]
@r3v4-onbloc did you check the docs on [Extending OpenAPI](https://fastapi.tiangolo.com/advanced/extending-openapi/?h=extend)?

**Comment by r3v4-onbloc** [comment]
@iudeen 
yes i did. i'm not sure if i have to use swagger_ui_parameters. (i'm looking swagger ui docs too)

what i want to do is add a query parameter that every endpoints uses as las query parameter. and choose value from very top of docs page.

for example in above picture
 'Servers' will be 'version'
 '/v1' will be ['1', '2', '3']

and if i chose '2', all of my endpoints request will be go to 'api.mine.com/${funcName}?version=2'

**Comment by ItsDevMonk** [comment]
Do we have the multi select option ?

**Comment by tiangolo** [comment]
The UI is provided by Swagger UI, FastAPI only generates the OpenAPI schema.

The UI options are only those available in Swagger UI. I don't think there's a way to add a top level drop-down, but as Swagger UI is React, maybe you could create your own custom component extending and including Swagger UI. Probably a complex task, but if you need that, maybe that's the way to go.

About multiple options, not sure how Swagger UI renders it, but you can declare a query parameter that takes a list of enums.

**Comment by ItsDevMonk** [comment]
@tiangolo . Thanks for your input. I tired with List of Enums, but with this option, dropdown option is not visible. But I can understand about the difficulties. Thanks again.

**Comment by tiangolo** [comment]
Thanks for reporting back @ItsDevMonk !

**Comment by CAGalbraith** [comment]
>Do we have the multi select option ?

It looks like you can select multiple values from a `list[Enum]` in the swagger UI by holding down control and selecting each value.
